### ISSUE TITLE
[no-misused-promises] should allow Promise<void> where void is expected (but not other promises)

### ISSUE THREAD

**Issue by jacekkopecky** [typescript-eslint/typescript-eslint#2979]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-misused-promises": [
      "error",
      {
        "checksVoidReturn": true
      }
    ]
  }
}
```

```TS
async function handler(): Promise<void> { /* */ }
document.body.addEventListener('click', handler);
```

tsconfig:

```json
{
  "compilerOptions": {
    "sourceMap": true,
    "strictNullChecks": true,
    "strict": true,
    "noImplicitAny": true,
    "target": "es2020",
    "module": "es2020",
    "allowJs": true,
    "esModuleInterop": true
  },
  "include": [
    "src"
  ]
}
```

**Expected Result**

I expect the rule to allow the code above, possibly with a new option. 

I understand the intent of the rule as checking that `Promise<boolean>` (anything non-void) should not be used where a function with void return is expected. 

However, a function that returns `Promise<void>` should be quite acceptable.

The option `checksVoidReturn`, when set to `false`, is much too permissive – it's either check everything or check nothing; I would like the middle ground of treating `Promise<void>` as void when checking return.

I use the `Promise<void>` function as an event handler, so nobody waits for the function to finish. In my experience, void-returning functions are normally passed as callbacks; the caller wouldn't wait for the function to finish.

But I understand that where a void-returning function is passed and the caller will need to wait for it to finish, treating `Promise<void>` as returning `void` would hide the asynchronous nature. That's why it probably wants to be an option.

I believe this is related to #1637 

**Actual Result**

The code above is rejected because `Promise<void>` doesn't match the expected `void`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.14.1` |
| `@typescript-eslint/parser`        | `4.14.1` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.18.0` |
| `node`                             | `14.7.0` |

**Comment by jacekkopecky** [comment]
I would agree with the linter to treat the cases differently there:
```ts
async () => someOpThatReturnsNonVoid()      // this one returns non-void
async () => { someOpThatReturnsNonVoid() }  // this one doesn't
```
Are you saying that many see the two cases above as semantically equivalent? I don't have that kind of experience with developers so really don't know – it's an honest question. But if someone wants either of the above to be used where `() => void` is expected, they would turn `checksVoidReturn` off anyway and my suggestion wouldn't affect them.

Same for your other case. If a function returns something, I'm happy for this rule to trigger (as long `checksVoidReturn` is true); but if it never returns anything, I'm suggesting the rule should be able to treat the following two as equivalent:
```ts
async () => { someOpThatReturnsNonVoid() }  // returns Promise<void>
() => { someOpThatReturnsNonVoid() }  // returns void
```

We are clearly focusing on different differences here.

**Comment by bradzacher** [comment]
What I'm saying is that I disagree with adding this to the rule for a few reasons:
1) `() => boolean` is assignable to `() => void`.
    Making a special case within this rule to allow `() => Promise<void>` to be assigned to `() => void`, but not `() => Promise<boolean>` is going to cause confusion for people.
2) In the context of `() => void` - `() => Promise.resolve(true)` is semantically equivalent to `() => Promise.resolve()` - either way the receiver is not going to consume the promise.
3) The step to remediate the lint error is to "make the function return `Promise<void>`. The easiest path to do this is to return nothing.. 

(1) is particularly troubling, because it creates asymmetry in the rule. Asymmetry can (and will confuse people), and when people are confused they'll either:
- add a confused eslint-disable comment (bad)
- disable the lint rule entirely (bad)
- file an issue or post on stackoverflow (bad)
- faff around and attempt to change their code until it works (bad)

I don't think that -- in the context of `no-misused-promises` -- this functionality is a good idea.

**Comment by jacekkopecky** [comment]
Ah, I wasn't aware that `() => boolean` is acceptable where `() => void` is expected. This changes my understanding of `checksVoidReturn`. Thank you.
